distributed/scheduler.py

```diff
@@ -7525,6 +7549,29 @@ def decide_worker(
             break
     else:
         ws = min(candidates, key=objective)
+
+    # If our group is large with few dependencies
+    # Then assign sequential tasks to similar workers, even if occupancy isn't ideal
+    if len(ts._group) > nthreads * 2 and sum(map(len, ts._group._dependencies)) < 5:
```

Comment:
* Isn't the length of all dependencies of a TG potentially very expensive? The length of a group iterates over all `TaskState`s in a given group. For some topologies, this would require us to iterate over all tasks (-1), wouldn't it?
* Is there any way to reason about the numeric values here? I think I'm still lacking intuition for TGs to tell how stable this heuristic is.

Reply:
> The length of a group iterates over all TaskStates in a given group

The code looks like this

```python
    def __init__(self, ...):
        self._states = {"memory": 0, "processing": 0, ...}

    def __len__(self):
        return sum(self._states.values())
```

So it's not as bad as it sounds.  However, iterating the dict of a few elements could still be concerning.  If so we could always keep a `_len` value around.  It would be cheap to maintain.

Reply:
> Is there any way to reason about the numeric values here? I think I'm still lacking intuition for TGs to tell how stable this heuristic is.

The 2 is because we want more than one task per worker to be allocated.  If there are more or equal workers as tasks then we're unlikely to co-schedule any tasks on similar workers, so this is a moot point.

The `< 5` is really saying "we want there to be almost no dependencies for the tasks in this group, but we're going to accept a common case of all tasks depending on some parameter or something like an abstract zarr file".  We're looking for cases where the dependency won't significantly affect the distribution of tasks throughout the cluster. This could be `len(dependencies) in (0, 1)` but we figured we'd allow a couple of these just in case. 

I expect that the distribution here will be bi-modal with tasks either in `(0, 1)` or in the hundreds or thousands.  Five seemed like a good separator value in that distribution.  I think that, given the distribution, this choice is stable and defensible.

Reply:
> The code looks like this

Right, that's `state` as in {Running, Memory, Released} and not state as in `TaskState` and is an aggregated dict. I was already a bit thrown off when I saw that. That's perfectly fine.

> I expect that the distribution here will be bi-modal with tasks either in (0, 1) or in the hundreds or thousands.

Thanks for the detailed description. I think I was thrown off by the TaskGroup semantics again. I was thinking about our typical tree reductions where we have usually *task* splits like 8 or 16. These are the situations where one would want to group all dependencies for the first reduction.
However, for *group* dependencies this should be a trivial dependency of *one*, correct?

Then, five is conservative, I agree 👍

Reply:
Perhaps `state` should have been called `state_counts`.  Oh well.

Ah, it's not `len(ts._group._dependencies)` which is what you're describing, I think.  It's `sum(map(len, ts._group._dependencies)) < 5`.  

We're counting up *all* of the dependencies for *all* of the tasks that are *like this task*.  So in a tree reduction, this number would likely be in the thousands for any non-trivially sized computation.  It is non-zero and less than five only in cases like the following:

```
a1 a2 a3 a4 a5 a6 a7 a8
 \  \  \  \  /  /  /  /
            b
```

Reply:
Really we're looking for cases where the number of dependencies, amortized over all similar tasks, is near-zero.

Reply:
This is the "ish" in "root-ish" tasks that we sometimes talk about here.

Reply:
> Perhaps state should have been called state_counts. Oh well.

naming is hard

> We're counting up all of the dependencies for all of the tasks that are like this task. So in a tree reduction, this number would likely be in the thousands for any non-trivially sized computation. It is non-zero and less than five only in cases like the following:
> Really we're looking for cases where the number of dependencies, amortized over all similar tasks, is near-zero.
> This is the "ish" in "root-ish" tasks that we sometimes talk about here.

I think I got it now. That's an interesting approach to gauge the local topology. What I'm currently wondering is if this or a closely related metric (e.g. ratio of group dependents/dependencies) could be used to estimate whether a task has the potential to increase/decrease parallelism. that'd be an interesting metric for work stealing.

anyhow, don't want to increase the scope here. this is a discussion we can delay. I'll let the professionals back to work! thanks!

Reply:
I think that it could be a useful metric for memory consuming/producing tasks.

It's also, yes, a good metric for increasing parallelism.  My experience though is that we are *always* in a state of abundant parallelism, and that scheduling to increase parallelism is not worth considering in our domain.  

Instead we should focus our scheduling decisions to reduce memory use and free intermediate tasks quickly.